src/DsaBottomUp.cc

```diff
@@ -50,10 +50,104 @@ static const Value *findUniqueReturnValue(const Function &F) {
   return onlyRetVal;
 }
 
+template <typename Set>
```

Comment:
I think this code computes marks on nodes. It would be more effective to have a few bits in the node data structure for marking. Instead of using an external set. 

Of course, in that case the marks have to be reset after use, but this is still just a single sweep over all nodes.